crates/bevy_ecs/src/schedule/condition.rs

```diff
@@ -80,6 +83,50 @@ pub mod common_conditions {
         }
     }
 
+    /// Generates a [`Condition`](super::Condition)-satisfying closure that returns `true`
+    /// if the resource of the given type has been added since the condition was last checked.
+    pub fn resource_added<T>() -> impl FnMut(Option<Res<T>>) -> bool
+    where
+        T: Resource,
+    {
+        move |res: Option<Res<T>>| match res {
+            Some(res) => res.is_added(),
+            None => false,
+        }
+    }
+
+    /// Generates a [`Condition`](super::Condition)-satisfying closure that returns `true`
+    /// if the resource of the given type has had its value changed since the condition was last checked.
+    pub fn resource_changed<T>() -> impl FnMut(Option<Res<T>>) -> bool
```

Comment:
Should this panic if the resource does not exist? The current behavior will be surprising, since removal won't count as changed.

Reply:
I think we should borrow the logic from resource_removed and add it to this.

Reply:
I don't think this should return true on resource removal. It would be surprising and not-user-friendly. It makes sense for this to behave exactly as our other change detection features do.

This is because, if you are using change detection, you probably care about doing something with the value whenever it changes. So it makes sense that it should only be true if the resource exists.

Removal does not count as changed with regular change detection on components and resources. It shouldn't here, either.

Reply:
Besides, if users care about detecting both "resource exists and the value changed" and "resource removed", they can trivially add both run conditions. They are composable.

If this run condition returns true on removal, there is no easy way to stop it from doing that.

Reply:
If we want to be really "complete" and pedantic, we could have:
 - `resource_changed`: true on `is_changed`, panic if resource does not exist
 - `resource_exists_and_changed`: true on `is_changed`, false if resource does not exist
 - `resource_changed_or_removed`: true on `is_changed`, true on removal (copy logic from `resource_removed`), false otherwise
 
 Let me know what you think.

Reply:
> if users care about detecting both "resource exists and the value changed" and "resource removed", they can trivially add both run conditions.

I agree with this stance now. We should leave it as is.

Reply:
> Besides, if users care about detecting both "resource exists and the value changed" and "resource removed", they can trivially add both run conditions

Surely we should make `resource_equals` work this way too, then.

Reply:
I agree that it makes sense (for the sake of API consistency) for `resource_equals` and `resource_changed` to behave the same way w.r.t missing resources and panicking.

Thing is, I don't know how I could implement an OR-combinator. If we don't have one, then we can't really go for the "composable" route right now.

If someone can tell me how to do it (so I can add it to this PR), or make a PR for it, then sure, let's do it! :) I will keep my `resource_changed` as is, and change `resource_equals` to work the same way.

Otherwise, maybe we should implement multiple separate standalone functions as I described earlier?

Reply:
>  I will keep my resource_changed as is, and change resource_equals to work the same way.

If we remove the panic from `resource_equals`, then any system using this run condition will just silently never run if the user forgets to initialize the resource. Counterintuitively, I think that panicking here improves the UX.

Reply:
Okay, then I think the least controversial path forward for now is to just make multiple separate functions, like I outlined earlier, so I'll do that. It will give users the flexibility and choice over what behavior they want. I'll keep the naming consistent with the other functions that already exist.

I suspected that the only way to make an OR-combinator might be using a custom System impl, and now you confirmed it. So, I think it's out of scope for this PR.

Reply:
Sounds good to me.

Reply:
Okay. I have implemented the variants we discussed above.

The naming should feel consistent with the resource-equality-checking RCs now

The behavior should feel consistent with Bevy change detection in general (and this is documented in the doc comments)

A changed-or-removed RC is available for those who want it.